accepted/0000-yomi-initial-integration.md

```diff
@@ -0,0 +1,361 @@
+SaltForm and a Salt state is 1:1, but using some empty `init.sls` we
+have the option of splitting a big SaltForm into multiple smaller
+SaltForms (subforms), that optionally belongs to the same group.
+
+The current `yomi-formula` is using this mechanism to split the pillar
+data into six sections:
+
+- yomi: general configuration data, like use `kexec` instead of
+  `reboot` after the installation, or some `systemd-firstboot` data.
+
+- yomi-bootloader: indicate the device where the bootloader will be
+  installed, and the parameters send to the kernel.
+
+- yomi-services: list of services that will be enabled, and is
+  `salt-minion` data will be copied into the new system.
+
+- yomi-software: repository registering and packages installation.
+
+- yomi-storage: device partition, LVM or RAID configuration, and file
+  systems applied.
+
+- yomi-users: creation of initial users, passwords and certificates.
+
+As we will see in future sections, those forms contains some
+overlapping with other salt-forms (the locale configuration that
+`systemd-firstboot` requires), and with other subsystems in Uyuni
+(like the software section).
+
+## Highstate and Workflow
+
+Installing a new system will be equivalent of applying a highstate to
+a system booted from the Yomi image.
+
+The current Yomi image is a JeOS based image with a `salt-minion`,
+some systemd overwrites that will help during the initial
+configuration, and the CLI tools required when a new installation is
+performed. You can find more information here:
+https://github.com/openSUSE/yomi#the-iso-image
+
+The current Yomi states are aware of differentiating when the system
+is booting from an external image (like the Yomi image), from when the
+boot comes from the local device. This is used to decide when the
+installation of the node makes sense or not when a highstate is
+re-applied.
+
+This means that we can set all the pillar data for the node, without
+risk of breaking or damaging already installed nodes when the Yomi
+state is in place for those systems.
+
+This open the question about the workflow that will be in place when a
+new system is allocated. The current prototype was tested like this
+one:
+
+1) Boot the new node via the Yomi ISO image. The local DHCP server
+   assign a FQDN to the system, and this is used automatically as a
+   minion ID.
+
+   (note: the master address was indicated via a Grub command line
+   that an overwrite in the image can read and interpret, how the new
+   minion will automatically discover the master is an open question
+   for me yet)
+
+2) Accept the Salt key in the manager node (Uyuni 4.0.2), and register
+   the system.
+
+3) Assign to the new node all the Yomi states (Installer group),
+   fulfill the SaltForms and apply the highstate to the new node.
+
+Once the node reboot, the highstate can be re-applied without
+consequences, as the state detect early that the root filesystem
+device is already mounted.
+
```

Comment:
See above. In a _perfectly_ idempotent world, applying the high state after changing something in the installer states _would_ re-run those states. But in the current design, the Salt minion and its tools would probably not survive if you re-partition the disk, right? So if we want to allow triggering the re-installation without having to PXE boot again, we'll probably always have to either use something like MicroOS (where we can apply the new OS to the snapshot) or boot/kexec into a RAM disk before the installer states are kicked off.

Reply:
> See above. In a _perfectly_ idempotent world, applying the high state after changing something in the installer states _would_ re-run those states. But in the current design, the Salt minion and its tools would probably not survive if you re-partition the disk, right?

Uhm. If you change the partition schema, but you do not boot using the PXE boot image or the USB image from Yomi, nothing will happen.

But if you boot from the Yomi image, the current implementation of the partitioner will detect a mismatch between the pillar and the current layout, and will refuse to remove anything.

The plan is to add a resize / wipefs option in the pillar in yomi, that will clean the devices or try to resize the partitions and the file system if this mismatch is present.

The wipefs state was implemented some time ago, but outside of Yomi to lower the risk of loosing data:

https://github.com/openSUSE/yomi/blob/master/README.md#cleaning-the-disks

 So if we want to allow triggering the re-installation without having to PXE boot again, we'll probably always have to either use something like MicroOS (where we can apply the new OS to the snapshot) or boot/kexec into a RAM disk before the installer states are kicked off.

Reply:
I will resolve this conversation, as Yomi is able to skip itself if the main device of the system is mounted, and the chroot cannot be done.

This skip the problem of breaking the system as a side effect of re-applying the highstate.